File: binary-tree-maxiumum-path.cc
#include <algorithm>
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
  int maxPathSum(TreeNode *root) {
    if (!root) return 0;
    int max_depth;
    return maxPath(root, &max_depth);
  }
private:
  // max_depth: max path down from the root.
  // NEGATIVE VALUES!
  int maxPath(TreeNode *root, int* max_depth) {
    int left_depth = 0, right_depth = 0;
    int left_max = 0, right_max = 0;
    if (root->left)
      left_max = maxPath(root->left, &left_depth);
    if (root->right)
      right_max = maxPath(root->right, &right_depth);
    *max_depth = max(0, max(left_depth, right_depth)) + root->val;
    int ret = root->val + max(0, left_depth) + max(0, right_depth);
    if (root->left)
      ret = max(ret, left_max);
    if (root->right)
      ret = max(ret, right_max);
    return ret;
  }
};

File: binary-tree-maxiumum-path.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {

    public int maxPathSum(TreeNode root) {
        return compute(root)[0];
    }

    /** return two values:
     *    1) all max path sum.
     *    2) max path sum from the root to a leaf.
     */
    private int[] compute(TreeNode root) {
        if (root == null) return new int[]{Integer.MIN_VALUE,
            Integer.MIN_VALUE};
        int[] leftSum = compute(root.left);
        int[] rightSum = compute(root.right);
        int[] sum = new int[2];
        sum[0] = Math.max(leftSum[0], rightSum[0]);
        int sumOverRoot = root.val + Math.max(0, leftSum[1]) +
            Math.max(0, rightSum[1]);
        if (sumOverRoot > sum[0]) sum[0] = sumOverRoot;
        sum[1] = root.val + Math.max(0, Math.max(leftSum[1], rightSum[1]));
        return sum;
    }
}
